File: javascript/cote/GJM/두니발_박사의_탈옥.mjs
import {readFile} from 'fs/promises'

const N = 50
const D = 100

function getDegree(adj) {
	const ret = new Array(adj.length).fill(0)

	for (let i = 0; i < adj.length; i++) {
		for (let j =0; j < adj.length; j++) {
			if (adj[i][j] != 0){
				ret[i]++;
			}
		}
	}

	return ret
}

function solve1(adj, query, d, prison) {
	const degrees = getDegree(adj)
	const path = [prison]

	function search(q) {
		let ret = 0
		const cur = path[path.length - 1]
		const curDay = path.length 

		if (curDay === d + 1) {
			if (cur != q) {
				return 0
			}

			ret = 1 

			for (let i = 0; i < path.length - 1; i++) {
				ret /= degrees[path[i]]
			}

			return ret
		}


		for (let i = 0; i < adj.length; i++) {
			if (adj[cur][i] != 1) {
				continue;
			}

			path.push(i)
			ret += search(q)
			path.pop()
		}

		return ret

	}

	const answer = []

	for (let q of query) {
		answer.push(search(q))
	}

	return answer
}

function solve2(adj, query, d, prison) {
	const degrees = getDegree(adj)
	let cache;

	const answer = []

	function search(cur, curDay, q) {
		if (curDay === d) {
			return (cur === q) ? 1. : 0.
		}

		let ret = cache[cur][curDay]

		if (ret > - 1) {
			return ret
		}
		ret = 0.

		for (let i = 0; i < adj.length; i++) {
			if (adj[cur][i] != 0) {
				ret += search(i, curDay + 1, q) / degrees[cur]
			}
		}
		cache[cur][curDay] = ret

		return ret
	}

	for (let q of query) {
		cache = new Array(adj.length + 1).fill(0)
		.map(_ => new Array(d + 1).fill(-1))
		answer.push(search(prison, 0, q))
	}	

	return answer
}

async function main() {
	const data = await readFile("data/두니발_박사의_탈옥.txt", {encoding : 'utf-8'})
		.then(lines => lines.split("\n").map(line => line.trim()))
	let cur = 0
	const testCases = parseInt(data[cur++])

	for (let i = 0; i < testCases; i++) {
		const [n, d, prison] =	data[cur++].split(" ")
			.map(i => parseInt(i))
		
		const adj = new Array(n).fill(0)
			.map(_ => data[cur++]. split(" ")
				.map(i => parseInt(i)))

		const query = data[cur++].split(" ")
			.map(i => parseInt(i))

		console.log(solve1(adj, query, d, prison), solve2(adj, query, d, prison));
	}
}

main()

File: c++/cote/GJM/두니발_박사의_탈옥.cpp
#include <fstream>
#include <iomanip>
#include <iostream>
#include <sstream>
#include <string>
#include <vector>

#define N 50
#define D 100

using namespace std;

double cache[N + 1][D + 1];

vector<int>& split(const string& s) {
  stringstream ss(s);
  string temp;
  vector<int>* v = new vector<int>();
  while (getline(ss, temp, ' ')) {
    v->push_back(stoi(temp));
  }
  return *v;
}

int** initialize(fstream& fd, int n) {
  string buf;
  int** map = new int*[n];

  for (int i = 0; i < n; i++) {
    map[i] = new int[n];
  }

  for (int i = 0; i < n; i++) {
    getline(fd, buf);
    vector<int> v = split(buf);
    for (int j = 0; j < n; j++) {
      map[i][j] = v[j];
    }
  }

  return map;
}

int* getDegree(int** map, int n) {
  int* deg = new int[n];
  memset(deg, 0, sizeof(*deg));

  for (int i = 0; i < n; i++) {
    for (int j = 0; j < n; j++) {
      if (map[i][j] != 0) deg[i]++;
    }
  }

  return deg;
}

// 그래프 dfs
double search1(int** map, int* deg, vector<int>& path, int n, int d, int q) {
  double ret = 0.;

  if (path.size() == d + 1) {
    if (path.back() != q) return 0.;

    ret = 1.;
    for (int i = 0; i + 1 < path.size(); i++) {
      ret /= deg[path[i]];
    }
    return ret;
  }

  int cur = path.back();

  for (int i = 0; i < n; i++) {
    if (map[cur][i] != 0) {
      path.push_back(i);
      ret += search1(map, deg, path, n, d, q);
      path.pop_back();
    }
  }

  return ret;
}

void solve1(int** map, int* deg, int n, int d, int p, vector<int>& query) {
  vector<int> path;
  path.push_back(p);

  for (auto q : query) {
    cout << fixed << setprecision(7) << search1(map, deg, path, n, d, q) << " ";
  }
  cout << endl;
}

double search2(int** map, int* deg, int cur, int days, int n, int d, int q) {
  if (days == d) {
    return (cur == q) ? 1. : 0.;
  }

  double& ret = cache[cur][days];
  if (ret > -0.5) return ret;

  ret = 0.;

  for (int i = 0; i < n; i++) {
    if (map[cur][i] != 0) {
      ret += search2(map, deg, i, days + 1, n, d, q) / deg[cur];
    }
  }

  return ret;
}

void solve2(int** map, int* deg, int n, int d, int p, vector<int>& query) {
  for (auto q : query) {
    memset(cache, -1, sizeof(cache));
    cout << fixed << setprecision(7) << search2(map, deg, p, 0, n, d, q) << " ";
  }
  cout << endl;
}

int main(void) {
  fstream fd("data/두니발_박사의_탈옥.txt");
  string buf;

  getline(fd, buf);
  int test_cases = stoi(buf);
  while (test_cases > 0) {
    getline(fd, buf);
    vector<int> v = split(buf);

    int** map = initialize(fd, v[0]);
    int* deg = getDegree(map, v[0]);
    getline(fd, buf);
    getline(fd, buf);
    vector<int> query = split(buf);
    solve1(map, deg, v[0], v[1], v[2], query);
    solve2(map, deg, v[0], v[1], v[2], query);

    for (int i = 0; i < v[0]; i++) {
      delete[] map[i];
    }
    delete[] map;

    delete[] deg;

    test_cases--;
  }

  return 0;
}